gateway/test/src/test/java/io/apiman/gateway/test/junit/vertx3/ApiToFileRegistry.java

```diff
@@ -0,0 +1,186 @@
+/*
+ * Copyright 2017 JBoss Inc
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.apiman.gateway.test.junit.vertx3;
+
+import io.apiman.gateway.engine.IEngineConfig;
+import io.apiman.gateway.engine.async.IAsyncResultHandler;
+import io.apiman.gateway.engine.beans.Api;
+import io.apiman.gateway.engine.beans.ApiContract;
+import io.apiman.gateway.engine.beans.Client;
+import io.apiman.gateway.engine.impl.InMemoryRegistry;
+import io.vertx.core.Vertx;
+import io.vertx.core.buffer.Buffer;
+import io.vertx.core.eventbus.EventBus;
+import io.vertx.core.file.FileSystem;
+import io.vertx.core.json.Json;
+import io.vertx.core.json.JsonArray;
+import io.vertx.core.json.JsonObject;
+
+import java.io.File;
+import java.io.IOException;
+import java.util.LinkedHashMap;
+import java.util.Map;
+
+@SuppressWarnings("nls")
+public class ApiToFileRegistry extends InMemoryRegistry {
+
+    private FileSystem fileSystem;
+    private File file;
+    private JsonArray clients = new JsonArray();
+    private JsonArray apis = new JsonArray();
+    private JsonObject root = new JsonObject(); private void linkRoot() {
+        root.put("clients", clients);
+        root.put("apis", apis);
+    }
+    private Map<Client, JsonObject> clientMap = new LinkedHashMap<>();
+    private Map<Api, JsonObject> apiMap = new LinkedHashMap<>();
+    private EventBus eb;
+
+    public ApiToFileRegistry(Vertx vertx, IEngineConfig foo, Map<String, String> config) {
+        super();
+        this.eb = vertx.eventBus();
+        this.fileSystem = vertx.fileSystem();
+        linkRoot();
+        createTempFile();
+        createResetListener();
+    }
+
+    private void createResetListener() {
+        eb.consumer(ApiToFileRegistry.class.getCanonicalName(), msg -> {
+           if (msg.headers().get("action").equals("reset")) {
```

Comment:
At the end of the test run, a message is sent to all `ApiToFileRegistry` instructing them to reset. It runs in a separate vert.x instance to the code under test to avoid interference.

I guess I could destroy and redeploy but this is simpler and was a result of some other attempts I had earlier (which worked all bar one test because they were hard reloading which caused metrics to be dropped).